addons/test_mail_full/tests/test_portal.py

```diff
@@ -145,6 +147,188 @@ def test_portal_share_comment(self):
         self.assertEqual(message.author_id, self.partner_2)
 
 
+@tagged('portal')
+class TestPortalFlow(MailCommon, HttpCase):
+    """Share a link by email to a customer without an account for viewing a record through the portal.
+
+    The tests consist in sending a mail related to a record to a customer and checking that the record can be viewed
+    through the embedded link:
+    - either in the backend if the user is connected and has the right to
+    - or in the portal otherwise
+    """
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.customer = cls.env['res.partner'].create({
+            'country_id': cls.env.ref('base.be').id,
+            'email': 'mdelvaux34@example.com',
+            'lang': 'en_US',
+            'mobile': '+32492131422',
+            'name': 'Mathias Delvaux',
+            'phone': '+3227411823',
+        })
+        cls.record_portal = cls.env['mail.test.portal'].create({
+            'name': 'Test Portal Record',
+            'partner_id': cls.customer.id,
+            'user_id': cls.user_admin.id,
+        })
+        cls.mail_template = cls.env['mail.template'].create({
+            'auto_delete': True,
+            'body_html': '<p>Hello <t t-out="object.partner_id.name"/>, your quotation is ready for review.</p>',
+            'email_from': '{{ (object.user_id.email_formatted or user.email_formatted) }}',
+            'model_id': cls.env.ref('test_mail_full.model_mail_test_portal').id,
+            'name': 'Quotation template',
+            'partner_to': '{{ object.partner_id.id }}',
+            'subject': 'Your quotation "{{ object.name }}"',
+        })
+        cls._create_portal_user()
+        for group_name, group_func, group_data in cls.record_portal.sudo()._notify_get_recipients_groups():
+            if group_name == 'portal_customer' and group_func(cls.customer):
+                cls.record_access_url = group_data['button_access']['url']
+                break
+        else:
+            raise AssertionError('Record access URL not found')
+        # Build record_access_url_wrong_token with altered access_token for testing security
+        parsed_url = url_parse(cls.record_access_url)
+        query_params = url_decode(parsed_url.query)
+        cls.record_access_url_wrong_token = url_unparse(
+            (parsed_url[0], parsed_url[1], parsed_url[2],
+             url_encode({**query_params,
+                         'access_token': query_params['access_token'].translate(
+                             str.maketrans('0123456789abcdef',
+                                           '9876543210fedcba'))},
+                        sort=True),
+             parsed_url[4]))
+
+    def _assert_URL(self, url, path, expected_fragment_params=None, expected_in_query=None):
```

Comment:
Would add a small docstring :) and maybe make it public (aka without underscore) ? Not sure we have to keep it private-like.

Reply:
Yes, good idea. By doing docstring helped me to make an easier method to understand.